animation_nodes/nodes/color/vertex_color_input.py

```diff
@@ -0,0 +1,101 @@
+import bpy
+from bpy.props import *
+from ... events import propertyChanged
+from ... data_structures import Color, ColorList, VirtualColorList
+from ... base_types import AnimationNode, VectorizedSocket
+
+modeItems = [
+    ("ALL", "All", "Get color of every vertex", "NONE", 0),
+    ("INDEX", "Index", "Get color of a specific vertex", "NONE", 1)
+]
+
+colorLayerIdentifierTypeItems = [
+    ("INDEX", "Index", "Get color layer based on the index", "NONE", 0),
+    ("NAME", "Name", "Get color layer based on the name", "NONE", 1)
+]
+
+class VertexColorInputNode(bpy.types.Node, AnimationNode):
```

Comment:
This should behave similar to the other node. There should be an anum to choose the mode (Loop, Vertex, or Polygon). The Vertex and Polygon outputs should be the average of the corresponding loops.

Reply:
Omar, I did not understand the _Vertex outputs should be the average of the loops_?

Reply:
When you are in Vertex mode, multiple loops will share the same vertex. In which case, the output color for each vertex should be the average of all vertices that share it. The same goes for polygons.

Reply:
Here is a code to get average colors for polygons,
```
sourceMesh = object.an.getMesh(False)
polygonIndices = sourceMesh.an.getPolygonIndices()

colorsList = self.getVertexColors(object, identifier)
polygonsColors = ColorList(length = len(polygonIndices))

index = 0 
for i in range(len(polygonIndices)):
    r = 0
    g = 0
    b = 0
    a = 0
    polyLength = polygonIndices.polyLengths[i]
    for j in range(polyLength):
        color = colorsList[index]
        r += color.r/polyLength
        b += color.b/polyLength
        g += color.g/polyLength
        a += color.a/polyLength
        index += 1
    polygonsColors[i] = Color((r, g, b, a))
```

Reply:
Sure. But it should also be implemented in Cython. You can add utility functions in `math/Color.pyx` if you need.

Reply:
Yes, after finalizing it, I'll implement in the Cython. I think it would be convenient if we have simple math operations for colors e.g., `colorA = ColorB + ColorC` or 'ColorA = ColorB/(any number)'.

Here is code for Vertex average,
```
sourceMesh = object.an.getMesh(False)
polygonIndices = sourceMesh.an.getPolygonIndices()
vertexCount = len(sourceMesh.vertices)

colorsList = self.getVertexColors(object, identifier)
vertexsColors = ColorList(length = vertexCount)

for i in range(vertexCount):
    r = 0
    g = 0
    b = 0
    a = 0
    shareCount = 0
    for j in range(len(polygonIndices.indices)):
        polyIndex = polygonIndices.indices[j]
        if polyIndex == i:
            color = colorsList[j]
            r += color.r
            b += color.b
            g += color.g
            a += color.a
            shareCount += 1
    vertexsColors[i] = Color((r/shareCount, g/shareCount, b/shareCount, a/shareCount))
return vertexsColors
```

Reply:
Omar, these are the functions for Cython code,

For vertex colors from loops colors:
```
def getVertexColorsFromLoopColors(long vertexCount, PolygonIndicesList polygons, VirtualColorList colors):
    cdef Color color
    cdef double r, g, b, a
    cdef long i, j, shareCount
    cdef ColorList vertexsColors = ColorList(length = vertexCount)
            
    for i in range(vertexCount):
        r = 0
        g = 0
        b = 0
        a = 0
        shareCount = 0
        for j in range(len(polygons.indices)):
            if polygons.indices[j] == i:
                color = colors.get(polygons.indices[j])[0]
                r += color.r
                b += color.b
                g += color.g
                a += color.a
                shareCount += 1
        vertexsColors.data[i] = Color(r/shareCount, g/shareCount, b/shareCount, a/shareCount)
    return vertexsColors

```
For polygons colors from loops colors:
```
def getPolygonColorsFromLoopColors(PolygonIndicesList polygons, VirtualColorList colors):
    cdef Color color
    cdef double r, g, b, a
    cdef long i, j, index, polyLength
    cdef polygonsCount = len(polygons)
    cdef ColorList polygonsColors = ColorList(length = polygonsCount)
    
    index = 0 
    for i in range(polygonsCount):
        r = 0
        g = 0
        b = 0
        a = 0
        polyLength = polygons.polyLengths[i]
        for j in range(polyLength):
            color = colors.get(polygons.indices[index])[0]
            r += color.r/polyLength
            b += color.b/polyLength
            g += color.g/polyLength
            a += color.a/polyLength
            index += 1
        polygonsColors.data[i] = Color(r, g, b, a)
    return polygonsColors
```

Reply:
The `getVertexColorsFromLoopColors` function looks inefficient. Maybe something like this would be more efficient:
```
verticesColors = a list of zero colors of length verticesCount
loopCounts = a list of zero integers of length verticesCount
for i in indices:
    verticesColors[i] += loopColors[i]
    loopCounts[i] += 1

for i in range(verticesCount):
    verticesColors[i] /= loopCounts[i]

return verticesColors
```

Reply:
Omar, How I can create ColorList with zero color `vertexsColors = ColorList(length = vertexCount)`?

Reply:
Ahh, I found the syntax to create  ColorList with zero color `vertexsColors = ColorList.fromValue(defaultColor, length = vertexCount)`, is it correct way?

Reply:
Here is code of `getVertexColorsFromLoopColors` function for Cython,
```
def getVertexColorsFromLoopColors(long vertexCount, PolygonIndicesList polygons, VirtualColorList colors):
    cdef long i, index
    cdef double r, g, b, a
    cdef Color loopColor, vertexColor
    cdef ColorList vertexsColors = ColorList.fromValue((0, 0, 0, 0), length = vertexCount)
    cdef LongList loopCounts = LongList.fromValue(0, length = vertexCount)        
    
    for i in range(len(polygons.indices)):
        index = polygons.indices[i]
        vertexColor = vertexsColors.data[index]
        loopColor = colors.get(polygons.indices[i])[0]
        vertexColor.r += loopColor.r
        vertexColor.b += loopColor.b
        vertexColor.g += loopColor.g
        vertexColor.a += loopColor.a
        vertexsColors.data[index] = vertexColor
        loopCounts[index] += 1

    for i in range(vertexCount):
        vertexColor = vertexsColors.data[i]
        vertexColor.r = vertexColor.r/loopCounts[i]
        vertexColor.b = vertexColor.b/loopCounts[i]
        vertexColor.g = vertexColor.g/loopCounts[i]
        vertexColor.a = vertexColor.a/loopCounts[i]
        vertexsColors.data[i] = vertexColor
    return vertexsColors
```

Reply:
Nice. Now add some utility functions in `math/color.pyx` to handle color addition and division.

Reply:
Okay. I have a question,
- What are the basic points I have to take into account while adding any utility function?

Here is the mockup of the code, 
```
from . color cimport Color

cdef addColor(Color colorA, Color colorB):
    cdef double r, g, b, a
    r = colorA.r + colorB.r
    g = colorA.g + colorB.g
    b = colorA.b + colorB.b
    a = colorA.a + colorB.a
    return Color(r, g, b, a)

cdef divideColor(Color color, double x):
    color.r = color.r/x
    color.g = color.g/x
    color.b = color.b/x
    color.a = color.a/x
    return color
```
It compiled but gave a warning `math/color.c:1154:18: warning: ‘__pyx_f_15animation_nodes_4math_5color_divideColor’ defined but not used [-Wunused-function]`

Reply:
- Notice that there is a `color.pxd` file. This is like a header file, where you put definitions and function prototypes.
- Functions that operates on structs (Like Color), should take pointers to structs as arguments, including the output. Then you should write the output to that color instead of returning it. For instance:
  ```cython
  cdef void addColor(Color* target, Color* a, Color* b):
    target.r = a.r + b.r
    target.g = a.g + b.g
    target.b = a.b + b.b
    target.a = a.a + b.a
  ```
- Some functions are said to operate "Inplace". Those functions write the computed values to the input directly. For instance:
  ```cython
  cdef void scaleColor_Inplace(Color* c, float factor):
    c.r *= factor
    c.g *= factor
    c.b *= factor
    c.a *= factor
  ```

The warning you attached will be gone as soon as you use the function.

Reply:
Thanks for the explanation. Is this how they will be defined in `color.pxd` file,
```
from . color cimport Color
cdef void addColor(Color target, Color a, Color b)
cdef void scaleColor_Inplace(Color c, float factor)
```

Reply:
Okay. I have found the correct way,
```
cdef struct Color:
    float r, g, b, a

cdef void addColor(Color* target, Color* a, Color* b)
cdef void scaleColor_Inplace(Color* c, float factor)
```
and then I call in the `color/c_utils.pyx` file,
```
from ... math cimport (
    addColor
)
```
But it gives an error,
```
File "animation_nodes/nodes/color/c_utils.pyx", line 1, in init animation_nodes.nodes.color.c_utils
AttributeError: module 'animation_nodes.math.color' has no attribute '__pyx_capi__'

```

Reply:
It is not immediately clear to me why this error occurs. Perhaps the code didn't compile correctly. Can you recompile?

Reply:
There is no error while compiling but when I run the blender and add the Vertex Input Node (have an object with a Col layer) then that error shows up. Here is the full screenshot, 
![Screenshot from 2019-11-30 00-31-14](https://user-images.githubusercontent.com/30294746/69886627-dd331500-1308-11ea-8db5-cb63f39f10c9.png)

Reply:
Can you commit the changes so that I can investigate this better?

Reply:
My bad. I just found the cause. Actually, I forgot to copy the newly compiled `color.cpython-37m-x86_64-linux-gnu.so` in `animation_nodes/math/` folder (using in the blender).

Reply:
You are not using the `--copy` flag?

Reply:
Now, I'm using that flag.

Omar, this function, can be used for `multiply` as well as `divide` color with a number,
```
cdef void scaleColor_Inplace(Color* c, float factor):
  c.r *= factor
  c.g *= factor
  c.b *= factor
  c.a *= factor
```
Questions:
- Should we need to create separate functions for `multiply` and `divide` operations or use the above function?
- If we are going to keep the above function then what will be its name, `scaleColor_Inplace` or `scaleColor` or `multipy_divide_Color` or sth else?

Reply:
Keep the name. Use it for division.

Reply:
Here are functions of `c_utils.pyx` file with `addColor` and `scaleColor_Inplace`,
```
def getVertexColorsFromLoopColors(long vertexCount, PolygonIndicesList polygons, VirtualColorList colors):
    cdef long i, index
    cdef Color vertexColor
    cdef ColorList vertexsColors = ColorList.fromValue((0, 0, 0, 0), length = vertexCount)
    cdef LongList loopCounts = LongList.fromValue(0, length = vertexCount)        
    
    for i in range(len(polygons.indices)):
        index = polygons.indices[i]
        vertexColor = vertexsColors.data[index]
        addColor(&vertexColor, &vertexColor, &colors.get(polygons.indices[i])[0])
        vertexsColors.data[index] = vertexColor
        loopCounts[index] += 1

    for i in range(vertexCount):
        scaleColor_Inplace(&vertexsColors.data[i], 1/loopCounts[i])
    return vertexsColors
```
and 

```
def getPolygonColorsFromLoopColors(PolygonIndicesList polygons, VirtualColorList colors):
    cdef Color color
    cdef long i, j, index, polyLength
    cdef polygonsCount = len(polygons)
    cdef ColorList polygonsColors = ColorList(length = polygonsCount)
    
    index = 0 
    for i in range(polygonsCount):
        color = Color(0, 0, 0, 0)
        polyLength = polygons.polyLengths[i]
        for j in range(polyLength):
            addColor(&color, &color, &colors.get(polygons.indices[index])[0])
            index += 1
        scaleColor_Inplace(&color, 1/polyLength)
        polygonsColors.data[i] = color
    return polygonsColors
```

Reply:
Ok, just commit directly and I will ad a review there.